src/bridge/AdminRoomHandler.ts

```diff
@@ -484,7 +484,7 @@ export class AdminRoomHandler {
                     `Username is longer than the maximum permitted by the bridge (${SANE_USERNAME_LENGTH}).`
                 );
             }
-            else if (IdentGenerator.validateUsername(username)) {
+            else if (!IdentGenerator.validateUsername(username)) {
```

Comment:
@Half-Shot what's the point of having a bug in the first commit which then is being fixed *(with no explanation btw)* in the latter commit, if you can just have no bug at all? Please squash the first and last commits into one.

Also, I'm not sure what are the project's rules, but I presume that the changelog entry *(whose commit title also would need fixing, because just `changelog` tells basically nothing about the commit)* is better to add at the same commit as the one that implemented functional, in which case it's better to squash with the other two as well.

So I'd recommend to squash all 3 commits into one.

Reply:
> what's the point of having a bug in the first commit which then is being fixed (with no explanation btw) in the latter commit, if you can just have no bug at all?

if you look at the timestamps, the last one was authored almost an hour later; so that's because the commits match the actual development history

Reply:
> > what's the point of having a bug in the first commit which then is being fixed (with no explanation btw) in the latter commit, if you can just have no bug at all?
> 
> if you look at the timestamps, the last one was authored almost an hour later; so that's because the commits match the actual development history

That's… an interesting fact, but this "history" shouldn't have gotten through to the PR. The obvious problem is this would cause bisectability problems. Then there's also that code review often starts with the first commit and this bug being there is just waste of time for a reviewer who finds it, types a comment about the problem and a possible fix, only to later find out that author did address the problem, but for some reason in another commit.

Reply:
> this "history" shouldn't have gotten through to the PR

the "Flip check" commit was authored after Half-Shot opened the PR. This is the correct way to amend pull-requests; because the alternative is to force-push every time, which has a much worse UI on GitHub (and similar forges), and lacks a description of what each change is.

Then the PR will squashed-merged, like every other PR in this repo.

Reply:
> This is the correct way to amend pull-requests; because the alternative is to force-push every time, which has a much worse UI on GitHub (and similar forges), and lacks a description of what each change is.

FTR, Github allows to see code changes between two force-pushes. If the problem is just lack of description attached to them, it may be worked around through administrative means, such as asking to pair every force-push with a comment what was done. Another way, which sometimes employed in other FOSS projects is adding a description of what changed in the body of the amended commit.

Creating new commit for each fix hurts stronger than the obliging contributors to elaborate what changed, because that makes problems to code review, which is usually done by maintainers. And maintainers' time is more precious than that of contributors.

This isn't just my opinion, that works like this in all large source projects: Linux kernel, Mesa, Mutter, libinput, X11… As a matter of fact, since this is a Matrix plugin, I opened a list of PRs for Element sorted them by most commented, and looked at a few random, there are also force-pushes, for example: https://github.com/vector-im/element-desktop/pull/170

Like, imagine you send a large refactoring that amounts to 40 patches to Linux kernel, and then get a feedback for most of them, like asking to add a comment or to make use of an existing function… So what, you address the feedback, and send back 70 patches, where first 40 ones still contain the issues, and then new 30 ones are the fixes to those older ones? 😂😂😂 And perhaps with a very informative title like [this one in the PR we discuss](https://github.com/matrix-org/matrix-appservice-irc/pull/1538/commits/5c9da9d7c7f0cdc3e27fe110c9c920f68b91079d)? 😄😄😄

> Then the PR will squashed-merged, like every other PR in this repo.

Wow… I confirm that, I looked through various `Contributing.md` files, and there's a note saying commits are always squashed. And permits to push new commits fixing old ones *(doesn't require doing so though)*  So, like, if I refactor the project creating 20 patches, there will be no sane history, upon merging it will turn into an indistinguishable wall of changes, without a way to revert a single commit or even understand what was done.

Oh, well. That's the policy I guess, no point to discuss it here then 🤷‍♂️

Sorry, I see no button "resolve", if somebody can close this discussion, please do so.

Reply:
> FTR, Github allows to see code changes between two force-pushes

I'm aware,that's why I said "worse UI", not "no UI"

> Creating new commit for each fix hurts stronger than the obliging contributors to elaborate what changed, because that makes problems to code review, which is usually done by maintainers

How so? they don't have to review commit-per-commit.

> Like, imagine you send a large refactoring that amounts to 40 patches to Linux kernel, and then get a feedback for most of them, like asking to add a comment or to make use of an existing function… So what, you address the feedback, and send back 70 patches, where first 40 ones still contain the issues, and then new 30 ones are the fixes to those older ones?  And perhaps with a very informative title like this one in the PR we discuss? 

Yeah, and that's why large projects like Linux don't (or shouldn't) use Github.

> , like, if I refactor the project creating 20 patches, there will be no sane history, upon merging it will turn into an indistinguishable wall of changes, without a way to revert a single commit or even understand what was done.

then don't send 20 patches in a single PR.

Reply:
> How so? they don't have to review commit-per-commit.

Sure, you can start review from the global PR diff. But for good code review you need to also check individual commits for problems, both in the code and in description *(because this will be the history that you and other contributors will return to whenever some code raises questions and you want to understand why it was added)*. So actually, one has to.

> Yeah, and that's why large projects like Linux don't (or shouldn't) use Github.

Well, Linux kernel aside, all other projects I mentioned use Gitlab and the force-push UI there is similar.

> then don't send 20 patches in a single PR.

Is there a better way? If you suggest sending each one in a separate PR, this will be a nightmare to review, because commits in a PR usually rely on the previous ones.

Reply:
> > How so? they don't have to review commit-per-commit.
> 
> Sure, you can start review from the global PR diff. But for good code review you need to also check individual commits for problems, both in the code and in description _(because this will be the history that you and other contributors will return to whenever some code raises questions and you want to understand why it was added)_. So actually, one has to.

Oh, but you are right of course in the context of `matrix-org` policy to squash everything. In that case no point to look at individual commits. My answer was more generic, as in, suitable for usual projects, who retain the original commit history.

Reply:
> Well, Linux kernel aside, all other projects I mentioned use Gitlab and the force-push UI there is similar.

I meant PR-based workflows, not Github specifically. This includes Gitlab and Gitea.

Personally I find patch-based workflows (by email like Linux, Phabricator, or Gerrit) much better for this, but alas they aren't popular anymore.

> Is there a better way? If you suggest sending each one in a separate PR, this will be a nightmare to review, because commits in a PR usually rely on the previous ones.

If you look at this repo, I have a bunch of open PRs that build toward a single feature I want, and all but the last one have independent commits. Yeah it's painful to write PRs this way, but it's the best way to keep them readable despite Github's limitations.